flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -284,19 +300,45 @@ private void cleanUpPathQuietly(@Nonnull Path path) {
     private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandles)
             throws Exception {
 
-        // Prepare for restore with rescaling
-        KeyedStateHandle initialHandle =
+        Preconditions.checkArgument(restoreStateHandles != null && !restoreStateHandles.isEmpty());
+
+        // Choose the best state handle for the initial DB
+        final KeyedStateHandle selectedInitialHandle =
                 RocksDBIncrementalCheckpointUtils.chooseTheBestStateHandleForInitial(
                         restoreStateHandles, keyGroupRange, overlapFractionThreshold);
 
-        // Init base DB instance
-        if (initialHandle != null) {
-            restoreStateHandles.remove(initialHandle);
-            initDBWithRescaling(initialHandle);
-        } else {
-            this.rocksHandle.openDB();
+        Preconditions.checkNotNull(selectedInitialHandle);
+
+        final Path absolutInstanceBasePath = instanceBasePath.getAbsoluteFile().toPath();
+
+        final List<StateHandleDownloadSpec> allDownloadSpecs =
+                new ArrayList<>(restoreStateHandles.size());
+
+        // Prepare and collect all the download request to pull remote state to a local directory
+        for (KeyedStateHandle stateHandle : restoreStateHandles) {
+            if (!(stateHandle instanceof IncrementalRemoteKeyedStateHandle)) {
+                throw unexpectedStateHandleException(
+                        IncrementalRemoteKeyedStateHandle.class, stateHandle.getClass());
+            }
+            StateHandleDownloadSpec downloadRequest =
+                    new StateHandleDownloadSpec(
+                            (IncrementalRemoteKeyedStateHandle) stateHandle,
+                            absolutInstanceBasePath.resolve(UUID.randomUUID().toString()));
+
+            if (stateHandle == selectedInitialHandle) {
+                // Insert the initial handle at the head of the list
+                allDownloadSpecs.add(0, downloadRequest);
+            } else {
+                allDownloadSpecs.add(downloadRequest);
+            }
```

Comment:
`add(0)`/`get(0)` seems unclear and fragile to me.
How about using map for `allDownloadSpecs` instead?

```
Map<KeyedStateHandle,StateHandleDownloadSpec> allDownloadSpecs = new HashMap<>();

...    

allDownloadSpecs.put(
    stateHandle,
    new StateHandleDownloadSpec(
        (IncrementalRemoteKeyedStateHandle) stateHandle,
        absolutInstanceBasePath.resolve(UUID.randomUUID().toString())));

...

initBaseDBForRescaling(allDownloadSpecs.get(selectedInitialHandle));
```